apps/aecontract/src/aect_call_tx.erl

```diff
@@ -182,19 +182,32 @@ process(#contract_call_tx{caller_id   = CallerId,
     %% the call object.
     {Call, Trees3} = run_contract(CallTx, Call0, Env, Trees2),
 
-    %% Refund unused gas.
+    %% Roll back state changes if call failed
     Trees4 =
+        case aect_call:return_type(Call) of
+            ok -> Trees3;
+            E  ->
+                lager:debug("Contract call error ~w ~w~n",[E, Call]),
+                %% We still need to charge gas and fee
+                case Charges of
+                    0 -> Trees1;
```

Comment:
This seems reverting nonce change in case of `CALL` instruction. Could you please confirm?

I expect impacts of this are:
* `CALL` caller nonce not bumped if call errs or reverts.
* Contract call object in state tree overwritten (hence initial ones lost) in case of multiple `CALL`s by same caller in same generation (or same microblock, or same contract create/call tx) if without further nonce bumps in between.
  * Question that follows from this point is whether contract call object is/shall be written in contract call tree only for outermost create/call tx.

Related doc:
* https://github.com/aeternity/protocol/blame/14db60b7db4d51618ebad539aaf6a0b40b72e633/contracts/contract_state_tree.md#L40-L42
* https://github.com/aeternity/protocol/blame/14db60b7db4d51618ebad539aaf6a0b40b72e633/contracts/contract_state_tree.md#L62-L68

Reply:
I agree that the nonce still needs to be bumped. The nonce bump was hidden in the spend transaction previously.

Reply:
Note that the nonce is bumped for the outermost call (`Charges > 0`). If the call fails we (with this PR) roll back all state changes, including storing the call objects for inner calls, so we should either bump nonces *and* keep call objects, or do neither (which is what this PR implements).

I think it would be really weird to keep call objects for inner calls of contract executions that ultimately fail. Semantically it shouldn't matter if a call fails before or after an inner call.